### PR TITLE
Buchsbaum frequency

### PR DESCRIPTION
**Pull request by svincena**
A Hack Week version of the code from the "frequency" breakout room that attempts to add the buchsbaum_frequency function for calculating said frequency in a two-ion species, multi-fluid, magnetized plasma. The frequency occurs at a point when the perpendicular dielectric tensor element goes to zero due to the 180 degree out of phase motion for the two ion fluids.

### CODE REVIEW

File: hack/Buchsbaum_frequency.py [hunk 1]
```diff
@@ -36,35 +47,50 @@ def buchsbaum_frequency(B: Unit('T'), n1: u.m**-3, n2: u.m**-3, particle1: Parti
...
+        ``particle`` is not of an appropriate type.
+    `ValueError`
+        If the magnetic field contains invalid values or particle cannot
+        be used to identify a particle or isotope.
+    Warns
+    -----
+    : `~astropy.units.UnitsWarning`
+        If units are not provided, SI units are assumed.
         Notes
         -----
```
Comment (namurphy) [inline hack/Buchsbaum_frequency.py:66]:
```suggestion
        If units are not provided, SI units are assumed.

    Notes
    -----
```

File: hack/Buchsbaum_frequency.py [hunk 2]
```diff
@@ -36,35 +47,50 @@ def buchsbaum_frequency(B: Unit('T'), n1: u.m**-3, n2: u.m**-3, particle1: Parti
...
 
     In a magnetized plasma, the presence of two ion species allows the perpendicular
     component of the cold-plasma dielectric coefficient
     math:: \epsilon_{\perp} to vanish at an angular frequency referred to as the
-    Buchsbaum frequency[1], also called the bi-ion frequency or ion-ion hybrid frequency[2].
+    Buchsbaum frequency:cite:p:`buchsbaum:1960`., also called the bi-ion frequency or ion-ion hybrid frequency:cite:p:`vincena:2013`.
```
Comment (namurphy) [inline hack/Buchsbaum_frequency.py:70]:
```suggestion
    Buchsbaum frequency :cite:p:`buchsbaum:1960`, also called the bi-ion frequency or ion-ion hybrid frequency :cite:p:`vincena:2013`.
```

**Review by namurphy — COMMENTED** [review]
Thank you for submitting this!  I have a few minor formatting suggestions.  Overall, this looks like a fantastic addition.

File: hack/Buchsbaum_frequency.py [hunk 3]
```diff
@@ -0,0 +1,70 @@
+import numpy as np
+import astropy.units as u
+from plasmapy.formulary.frequencies import gyrofrequency, plasma_frequency
+from plasmapy.particles import Particle
+
+
+def buchsbaum_frequency(B: Unit('T'), n1: u.m**-3, n2: u.m**-3, particle1: Particle, particle2: Particle, Z1=None,
```
[the remark this answers is not in the document]
Comment (svincena) [reply hack/Buchsbaum_frequency.py]: This was fixed in a later commit.

File: hack/Buchsbaum_frequency.py [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+    Z2 : `float` or `~astropy.units.Quantity`, optional
+        The charge state for ion species #2. If not provided, it defaults to the charge number of ``particle2``.
+
+    Returns
+    -------
+    omega_BB : `~astropy.units.Quantity`
+        The Buchsbaum frequency of the plasma in units of radians per second.
+
+        Notes
+        -----
```
[the remark this answers is not in the document]
Comment (svincena) [reply hack/Buchsbaum_frequency.py]: Also fixed later.  Guess I made the PR to soon!

File: hack/Buchsbaum_frequency.py (hunk 2 above)
Comment (svincena) [reply hack/Buchsbaum_frequency.py:70]: Also fixed later. Guess I made the PR to soon!

File: hack/Buchsbaum_frequency.py [hunk 5]
```diff
@@ -0,0 +1,95 @@
+import numpy as np
+import astropy.units as u
```
Comment (rocco8773) [inline hack/Buchsbaum_frequency.py:5]:
```suggestion
__all__ = ["buchsbaum_frequency"]
__aliases__ = ["omega_bb_", "omega_ii_", "omega_bi_"]

import numpy as np
import astropy.units as u
```

File: hack/Buchsbaum_frequency.py [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+from plasmapy.formulary.frequencies import gyrofrequency, plasma_frequency
+from plasmapy.particles import Particle
+from plasmapy.utils.decorators import (
+    angular_freq_to_hz,
+    validate_quantities,
+)
```
Comment (rocco8773) [inline hack/Buchsbaum_frequency.py:12]:
```suggestion
)

__all__ += __aliases__
```

**Review by rocco8773 — CHANGES_REQUESTED** [review]
Define `__all__` and `__aliases__` in preparation for documentation.

File: hack/Buchsbaum_frequency.py (hunk 6 above)
Comment (svincena) [reply hack/Buchsbaum_frequency.py:12]: Thanks!

**Review by rocco8773 — CHANGES_REQUESTED** [review]
Now you need to add a `hack.Buchsbaum_frequency.rst` file to `/docs/api_static/` with the following content...

```rst
:orphan:

`hack.Buchsbaum_frequency`
==========================

.. currentmodule:: hack.Buchsbaum_frequency

.. automodapi:: hack.Buchsbaum_frequency

```

**Review by svincena — COMMENTED** [review]
Made changes per Erik and Nick

File: hack/Buchsbaum_frequency.py [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+    return np.sqrt((omega_p1**2 * omega_c2**2 + omega_p2**2 * omega_c1**2) / ( omega_p1**2 + omega_p2**2))
+
+omega_bb_ = buchsbaum_frequency
+"""Alias to `~plasmapy.formulary.frequencies.buchsbaum_frequency`."""
+
+omega_ii_ = buchsbaum_frequency
+"""Alias to `~plasmapy.formulary.frequencies.buchsbaum_frequency`."""
+
+omega_bi_ = buchsbaum_frequency
+"""Alias to `~plasmapy.formulary.frequencies.buchsbaum_frequency`."""
```
Comment (rocco8773) [inline hack/Buchsbaum_frequency.py]:
```suggestion
o_bb_ = buchsbaum_frequency
"""Alias to `~plasmapy.formulary.frequencies.buchsbaum_frequency`."""

o_ii_ = buchsbaum_frequency
"""Alias to `~plasmapy.formulary.frequencies.buchsbaum_frequency`."""

o_bi_ = buchsbaum_frequency
"""Alias to `~plasmapy.formulary.frequencies.buchsbaum_frequency`."""
```

The miss-match naming to `__aliases__` is what's causing the doc build to currently fail.

Comment (svincena) [reply hack/Buchsbaum_frequency.py]: done. thanks.

File: hack/Buchsbaum_frequency.py [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+    }
+)
+@angular_freq_to_hz
+def buchsbaum_frequency(B: u.T, n1: u.m**-3, n2: u.m**-3, particle1: Particle, particle2: Particle,
+                        Z1=None, Z2=None) -> u.rad / u.s:
+
+    r"""
+    Calculate the Buchsbaum frequency in units of radians per second.
+    **Aliases:** `omega_bb_`, `omega_ii_`, `omega_bi_`
+    Parameters
```
Comment (rocco8773) [inline hack/Buchsbaum_frequency.py]:
```suggestion
    Calculate the Buchsbaum frequency in units of radians per second.
    
    **Aliases:** `o_bb_`, `o_ii_`, `o_bi_`
    
    Parameters
```

We need spaces here so `sphinx` can properly parse, and the aliases need to be updated.

File: hack/Buchsbaum_frequency.py [hunk 9]
```diff
@@ -0,0 +1,100 @@
...
+    Z2 : `float` or `~astropy.units.Quantity`, optional
+        The charge state for ion species #2. If not provided, it defaults to the charge number of ``particle2``.
+
+    Returns
+    -------
+    omega_BB : `~astropy.units.Quantity`
+        The Buchsbaum frequency of the plasma in units of radians per second.
+        Setting keyword ``to_hz=True`` will apply the factor of :math:`1/2π`
+        and yield a value in Hz.
+    Raises
```
Comment (rocco8773) [inline hack/Buchsbaum_frequency.py]:
```suggestion
        and yield a value in Hz.
    
    Raises
```

File: hack/Buchsbaum_frequency.py [hunk 10]
```diff
@@ -0,0 +1,100 @@
...
+    ------
+    `TypeError`
+        If the magnetic field is not a `~astropy.units.Quantity` or
+        ``particle`` is not of an appropriate type.
+    `ValueError`
+        If the magnetic field contains invalid values or particle cannot
+        be used to identify a particle or isotope.
+    Warns
```
Comment (rocco8773) [inline hack/Buchsbaum_frequency.py]:
```suggestion
        be used to identify a particle or isotope.

    Warns
```

File: hack/Buchsbaum_frequency.py [hunk 11]
```diff
@@ -0,0 +1,100 @@
...
+    math:: \epsilon_{\perp} to vanish at an angular frequency referred to as the
+    Buchsbaum frequency:cite:p:`buchsbaum:1960`., also called the bi-ion frequency or ion-ion hybrid frequency:cite:p:`vincena:2013`.
+    This frequency can be defined as:
+    .. math::
+    \omega_{BB} \equiv \sqrt{\frac{\omega_{p1}^{2}\omega_{c2}^{2} + \omega_{p2}^{2}\omega_{c1}^{2}}{\omega_{p2}^{2}+\omega_{p2}^{2}}}
+
+#   These are the references, but do they need to go into a global bibtex-type document?
+#   [1] S. J. Buchsbaum, Phys. Fluids 3, 418 (1960); Resonance in a Plasma with Two Ion Species; https://doi.org/10.1063/1.1706052
+#   [2] S. T. Vincena, W. A. Farmer, J. E. Maggs, and G. J. Morales, Phys. Plasmas, 20, 012111 (2013);
+#    Investigation of an ion-ion hybrid Alfvén wave resonator https://doi.org/10.1063/1.4775777
```
Comment (rocco8773) [inline hack/Buchsbaum_frequency.py]: move these outside the docstring for now

**Review by rocco8773 — CHANGES_REQUESTED** [review]

File: hack/Buchsbaum_frequency.py (hunk 11 above)
Comment (svincena) [reply hack/Buchsbaum_frequency.py]: will do

File: hack/Buchsbaum_frequency.py [hunk 12]
```diff
@@ -0,0 +1,107 @@
+import numpy as np
+import astropy.units as u
+from plasmapy.formulary.frequencies import gyrofrequency, plasma_frequency
+from plasmapy.particles import Particle
+from plasmapy.utils.decorators import (
+    angular_freq_to_hz,
+    validate_quantities,
+)
+__all__ = ["buchsbaum_frequency"]
+__aliases__ = ["w_bb_", "w_ii_", "w_bi_"]
```
Comment (rocco8773) [inline hack/Buchsbaum_frequency.py]:
```suggestion
```

File: hack/Buchsbaum_frequency.py [hunk 13]
```diff
@@ -0,0 +1,107 @@
+import numpy as np
```
Comment (rocco8773) [inline hack/Buchsbaum_frequency.py:4]:
```suggestion
__all__ = ["buchsbaum_frequency"]
__aliases__ = ["w_bb_", "w_ii_", "w_bi_"]

import numpy as np
```

**Review by rocco8773 — CHANGES_REQUESTED** [review]
In practice, the `__all__` dunder is supposed to be defined at the top of the module, just after the module docstring.

File: hack/Buchsbaum_frequency.py [hunk 14]
```diff
@@ -0,0 +1,107 @@
...
+from plasmapy.particles import Particle
+from plasmapy.utils.decorators import (
+    angular_freq_to_hz,
+    validate_quantities,
+)
+__all__ = ["buchsbaum_frequency"]
+__aliases__ = ["w_bb_", "w_ii_", "w_bi_"]
+__all__ += __aliases__
+
+@validate_quantities(
```
Comment (rocco8773) [inline hack/Buchsbaum_frequency.py]:
```suggestion
__all__ += __aliases__


@validate_quantities(
```

2 blank lines before a function or class definition

File: hack/Buchsbaum_frequency.py [hunk 15]
```diff
@@ -0,0 +1,107 @@
...
+    validations_on_return={
+        "units": [u.rad / u.s, u.Hz],
+        "equivalencies": [(u.cy / u.s, u.Hz)],
+    }
+)
+@angular_freq_to_hz
+def buchsbaum_frequency(B: u.T, n1: u.m**-3, n2: u.m**-3, particle1: Particle, particle2: Particle,
+                        Z1=None, Z2=None) -> u.rad / u.s:
+
+    r"""
```
Comment (rocco8773) [inline hack/Buchsbaum_frequency.py]:
```suggestion
def buchsbaum_frequency(
    B: u.T,
    n1: u.m**-3,
    n2: u.m**-3,
    particle1: Particle,
    particle2: Particle,
    Z1=None,
    Z2=None,
) -> u.rad / u.s:
    r"""
```

File: hack/Buchsbaum_frequency.py [hunk 16]
```diff
@@ -0,0 +1,107 @@
...
+    Calculate the Buchsbaum frequency in units of radians per second.
+
+   **Aliases:** `w_bb_`, `w_ii_`, `w_bi_`
```
Comment (rocco8773) [inline hack/Buchsbaum_frequency.py]:
```suggestion
    **Aliases:** `w_bb_`, `w_ii_`, `w_bi_`
```

File: hack/Buchsbaum_frequency.py [hunk 17]
```diff
@@ -0,0 +1,107 @@
...
+        The charge state for ion species #2. If not provided, it defaults to the charge number of ``particle2``.
+
+    Returns
+    -------
+    omega_BB : `~astropy.units.Quantity`
+        The Buchsbaum frequency of the plasma in units of radians per second.
+        Setting keyword ``to_hz=True`` will apply the factor of :math:`1/2π`
+        and yield a value in Hz.
+    
+    Raises
```
Comment (rocco8773) [inline hack/Buchsbaum_frequency.py]:
```suggestion
        and yield a value in Hz.

    Raises
```

File: hack/Buchsbaum_frequency.py [hunk 18]
```diff
@@ -0,0 +1,115 @@
...
+    : `~astropy.units.UnitsWarning`
+        If units are not provided, SI units are assumed.
+
+    Notes
+    -----
+    In a magnetized plasma, the presence of two ion species allows the perpendicular
+    component of the cold-plasma dielectric coefficient
+    math:: \epsilon_{\perp} to vanish at an angular frequency referred to as the
+    Buchsbaum frequency:cite:p:`buchsbaum:1960`., also called the bi-ion frequency or ion-ion hybrid frequency:cite:p:`vincena:2013`.
+    This frequency can be defined as:
```
Comment (rocco8773) [inline hack/Buchsbaum_frequency.py]:
```suggestion
    In a magnetized plasma, the presence of two ion species allows the
    perpendicular component of the cold-plasma dielectric coefficient
    :math:`\epsilon_{\perp}` to vanish at an angular frequency referred
    to as the Buchsbaum frequency :cite:p:`buchsbaum:1960`, also called
    the bi-ion frequency or ion-ion hybrid frequency
    :cite:p:`vincena:2013`.  This frequency can be defined as:
```

File: hack/Buchsbaum_frequency.py [hunk 19]
```diff
@@ -0,0 +1,115 @@
...
+
+    .. math::
+    \omega_{BB} \equiv \sqrt{\frac{\omega_{p1}^{2}\omega_{c2}^{2} + \omega_{p2}^{2}\omega_{c1}^{2}}{\omega_{p2}^{2}+\omega_{p2}^{2}}}
```
Comment (rocco8773) [inline hack/Buchsbaum_frequency.py]:
```suggestion
    .. math::
        \omega_{BB} \equiv \sqrt{\frac{\omega_{p1}^{2}\omega_{c2}^{2}
            + \omega_{p2}^{2}\omega_{c1}^{2}}{\omega_{p2}^{2}+\omega_{p2}^{2}}}
```

File: hack/Buchsbaum_frequency.py [hunk 20]
```diff
@@ -0,0 +1,110 @@
+__all__ = ["buchsbaum_frequency"]
+__aliases__ = ["w_bb_", "w_ii_", "w_bi_"]
+
+import numpy as np
+import astropy.units as u
+from plasmapy.formulary.frequencies import gyrofrequency, plasma_frequency
```
Comment (rocco8773) [inline hack/Buchsbaum_frequency.py]:
```suggestion
import astropy.units as u

from plasmapy.formulary.frequencies import gyrofrequency, plasma_frequency
```

File: hack/Buchsbaum_frequency.py [hunk 21]
```diff
@@ -0,0 +1,112 @@
...
+    omega_c1 = gyrofrequency(B, particle1, signed=False, Z=Z1)
+    omega_c2 = gyrofrequency(B, particle2, signed=False, Z=Z2)
+    omega_p1 = plasma_frequency(n1, particle1, z_mean=Z1)
+    omega_p2 = plasma_frequency(n2, particle2, z_mean=Z2)
+
+    return np.sqrt((omega_p1**2 * omega_c2**2 + omega_p2**2 * omega_c1**2) / ( omega_p1**2 + omega_p2**2))
+
+
+w_bb_ = buchsbaum_frequency
+"""Alias to `~plasmapy.formulary.frequencies.buchsbaum_frequency`."""
```
Comment (rocco8773) [inline hack/Buchsbaum_frequency.py]:
```suggestion
"""Alias to `~hack.Buchsbaum_frequency.buchsbaum_frequency`."""
```

**Review by rocco8773 — APPROVED** [review]

**Pull request merged**